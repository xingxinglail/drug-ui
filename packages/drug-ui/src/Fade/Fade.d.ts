import * as React from 'react';
import { TransitionProps } from 'react-transition-group/Transition';
import { SimpleSpread } from '..';
interface PropsExtra {
    timeout?: number | {
        enter?: number;
        exit?: number;
    };
}
export interface FadeProps extends SimpleSpread<TransitionProps, PropsExtra> {
}
export declare const name = "Fade";
declare const Fade: React.FC<FadeProps>;
export default Fade;
