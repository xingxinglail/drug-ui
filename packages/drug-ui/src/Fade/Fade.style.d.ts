import { Theme } from '../styles';
export declare const styles: (_: Theme) => Record<string, any>;
