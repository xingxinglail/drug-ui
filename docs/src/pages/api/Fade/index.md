---
filename: /packages/drug-ui/src/Fade/Fade.js
---

# Fade API

<p class="description">Fade React 组件的 API 文档。</p>

## Import

```js
import Fade from '@drug-ui/core/Fade';
// or
import { Fade } from '@drug-ui/core';
```

## Props

| Name | Type | Default | Description |
|:-----|:-----|:--------|:------------|
| <span class="prop-name">children</span> | <span class="prop-type">node</span> |  | 内容节点。 |
| <span class="prop-name">in</span> | <span class="prop-type">bool</span> | <span class="prop-default">false</span> | 如果为```true```，组件将开始进入过渡。 |
| <span class="prop-name">timeout</span> | <span class="prop-type">number &#124; { enter?: number, exit?: number }</span> | <span class="prop-default">{ enter: 225, exit: 195 }</span> | 过渡时长。 |

`ref` 被转发到根元素.

## CSS

Style sheet name: `DuiFade`.

## Demos

- [Transitions](/drug-ui/components/transitions/)




