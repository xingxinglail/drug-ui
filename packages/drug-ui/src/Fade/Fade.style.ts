import { Style, Styles } from 'jss';
import { Theme } from '../styles';

const container: Style = {
    opacity: 0,
    transition: 'opacity .3s cubic-bezier(0.4, 0, 0.2, 1)'
};

const entered: Style = {
    opacity: 1
};

const hidden: Style = {
    visibility: 'hidden'
};

export const styles = (_: Theme): Styles => ({
    container,
    entered,
    hidden
});
