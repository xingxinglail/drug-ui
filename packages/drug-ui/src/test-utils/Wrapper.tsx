import * as React from 'react';
import { ThemeProvider } from '../index';

interface Props {
    children?: any;
}

const Wrapper: React.FC<Props> = (props) => {
    return (
        <ThemeProvider>
            { props.children }
        </ThemeProvider>
    );
};

export default Wrapper;
