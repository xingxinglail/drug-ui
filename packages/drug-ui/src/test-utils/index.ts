export { default as Wrapper } from './Wrapper';
export { getClasses } from './getClasses';

export const wait = (delay = 50) => {
    return new Promise(resolve => setTimeout(resolve, delay));
};
