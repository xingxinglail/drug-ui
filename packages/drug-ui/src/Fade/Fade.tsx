import * as React from 'react';
import { Transition } from 'react-transition-group';
import { TransitionProps } from 'react-transition-group/Transition';
import classnames from 'classnames';
import { styles } from './Fade.style';
import { createUseStyles } from '../styles';
import { SimpleSpread } from '..';

interface PropsExtra {
    timeout?: number | { enter?: number, exit?: number };
}

export interface FadeProps extends SimpleSpread<TransitionProps, PropsExtra> {
}

export const name = 'Fade';

type FadeClassProps = 'container' | 'entered' | 'hidden';

const useStyles = createUseStyles<FadeClassProps>(styles, name);

const Fade: React.FC<FadeProps> = React.forwardRef<HTMLDivElement, FadeProps>((props, ref) => {
    const {
        className, style, in: inProp = false, timeout = {
            enter: 225,
            exit: 195
        }, onEnter, onEntering, onEntered, onExit, onExiting, onExited, children, ...rest
    } = props;
    const classes = useStyles();

    const handleTransitionEvent = (event: string, node: HTMLElement, isAppearing?: boolean) => {
        if (event === 'onEnter') {
            node.style.opacity = '0';
            onEnter && onEnter(node, isAppearing);
        }

        if (event === 'onEntering') {
            node.scrollTop;
            node.style.opacity = '1';
            node.style.transitionDuration = `${ typeof timeout === 'number' ? timeout : timeout.enter }ms`;
            onEntering && onEntering(node, isAppearing);
        }

        if (event === 'onEntered') {
            onEntered && onEntered(node, isAppearing);
        }

        if (event === 'onExit') {
            node.style.opacity = '1';
            onExit && onExit(node);
        }

        if (event === 'onExiting') {
            node.style.opacity = '0';
            node.style.transitionDuration = `${ typeof timeout === 'number' ? timeout : timeout.exit }ms`;
            onExiting && onExiting(node);
        }

        if (event === 'onExited') {
            onExited && onExited(node);
        }
    };

    return (
        <Transition
            in={ inProp }
            timeout={ timeout }
            onEnter={ (node, isAppearing) => handleTransitionEvent('onEnter', node, isAppearing) }
            onEntering={ (node, isAppearing) => handleTransitionEvent('onEntering', node, isAppearing) }
            onEntered={ (node, isAppearing) => handleTransitionEvent('onEntered', node, isAppearing) }
            onExit={ (node) => handleTransitionEvent('onExit', node) }
            onExiting={ (node) => handleTransitionEvent('onExiting', node) }
            onExited={ (node) => handleTransitionEvent('onExited', node) }
            { ...rest }>
            {
                state => (
                    <div
                        ref={ ref }
                        className={
                            classnames(
                                classes.container,
                                { [classes.entered]: state === 'entered' },
                                { [classes.hidden]: state === 'exited' && !inProp },
                                className
                            )
                        }
                        style={ { ...style } }>
                        { children }
                    </div>
                )
            }
        </Transition>
    );
});

Fade.displayName = name;

export default Fade;
