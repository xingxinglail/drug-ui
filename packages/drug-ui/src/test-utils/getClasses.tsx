import * as React from 'react';
import * as enzyme from 'enzyme';
import Wrapper from './Wrapper';

export const getClasses = (element: React.ReactElement) => {
    // @ts-ignore
    const { useStyles } = element.type;
    let classes;

    function Listener () {
        classes = useStyles();
        return null;
    }

    enzyme.mount(<Wrapper><Listener /></Wrapper>);
    return classes;
};
