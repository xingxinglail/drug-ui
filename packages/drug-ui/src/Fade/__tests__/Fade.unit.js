import * as renderer from 'react-test-renderer';
import * as React from 'react';
import * as enzyme from 'enzyme';
import Fade from '..';
import { Wrapper, getClasses, wait } from '@drug-ui/core/test-utils';

const mount = enzyme.mount;

describe('<Fade />', () => {

    let classes;

    it('render successful', () => {
        const json = renderer.create(<Wrapper><Fade /></Wrapper>).toJSON();
        expect(json).toMatchSnapshot();
        classes = getClasses(<Fade />);
    });

    it('默认渲染', () => {
        const wrapper = mount(<Wrapper><Fade>Children</Fade></Wrapper>);
        const classList = wrapper.getDOMNode().classList;
        expect(classList.contains(classes.container)).toBeTruthy();
        expect(classList.contains(classes.hidden)).toBeTruthy();
    });

    it('默认显示渲染', () => {
        const wrapper = mount(<Wrapper><Fade in>Children</Fade></Wrapper>);
        const classList = wrapper.getDOMNode().classList;
        expect(classList.contains(classes.container)).toBeTruthy();
        expect(classList.contains(classes.entered)).toBeTruthy();
    });

    it('触发所有钩子', (done) => {
        const onEnter = jest.fn();
        const onEntering = jest.fn();
        const onEntered = jest.fn(async () => {
            expect(window.getComputedStyle(dom).getPropertyValue('opacity')).toEqual('1');
            await wait();
            wrapper.setProps({ in: false });
        });
        const onExit = jest.fn();
        const onExiting = jest.fn();
        const onExited = jest.fn(() => {
            expect(onEnter).toHaveBeenCalledBefore(onEntering);
            expect(onEntering).toHaveBeenCalledBefore(onEntered);
            expect(onEntered).toHaveBeenCalledBefore(onExit);
            expect(onExit).toHaveBeenCalledBefore(onExiting);
            expect(onExiting).toHaveBeenCalledBefore(onExited);
            done();
        });

        const wrapper = mount(
            <Wrapper>
                <Fade
                    onEnter={ onEnter }
                    onEntering={ onEntering }
                    onEntered={ onEntered }
                    onExit={ onExit }
                    onExiting={ onExiting }
                    onExited={ onExited }>
                    <div />
                </Fade>
            </Wrapper>
        );
        const dom = wrapper.getDOMNode();
        expect(window.getComputedStyle(dom).getPropertyValue('opacity')).toEqual('0');
        wrapper.setProps({ in: true });
    });
});
