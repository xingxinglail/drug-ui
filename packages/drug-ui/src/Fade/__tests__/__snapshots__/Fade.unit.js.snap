// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Fade /> render successful 1`] = `
<div
  className="DuiFade-container DuiFade-hidden"
  style={Object {}}
>
  <div
    className="DuiFade-container DuiFade-hidden"
    style={Object {}}
  />
</div>
`;
